src/awkward/_v2/_connect/jax/nplike.py

```diff
@@ -0,0 +1,170 @@
+                ak.layout.Record,
+                ak.layout.ArrayBuilder,
+                ak.layout.LayoutBuilder32,
+                ak.layout.LayoutBuilder64,
+            ),
+        ):
+            return ak.operations.convert.to_jax(array, *args, **kwargs)
+
+        elif isinstance(array, Iterable):
+            return [self.to_rectilinear(x, *args, **kwargs) for x in array]
+
+        else:
+            ak._v2._util.error(ValueError("to_rectilinear argument must be iterable"))
+
+    def __getitem__(self, name_and_types):
+        ak._v2._util.error(
+            ValueError("__getitem__ for JAX Kernels is not implemented yet")
+        )
+
+    def __init__(self):
+        from awkward._v2._connect.jax import import_jax  # noqa: F401
+
+        self._module = import_jax().numpy
+
+    @property
+    def ma(self):
+        ak._v2._util.error(
+            ValueError(
+                "JAX arrays cannot have missing values until JAX implements "
+                "numpy.ma.MaskedArray" + ak._util.exception_suffix(__file__)
+            )
+        )
+
+    @property
+    def char(self):
+        ak._v2._util.error(
+            ValueError(
+                "JAX arrays cannot do string manipulations until JAX implements "
+                "numpy.char"
+            )
+        )
+
+    @property
+    def ndarray(self):
+        return self._module.ndarray
+
+    def asarray(self, array, dtype=None, order=None):
+        return self._module.asarray(array, dtype=dtype, order="K")
+
+    def raw(self, array, nplike):
+        if isinstance(nplike, Jax):
+            return array
+        elif isinstance(nplike, ak.nplike.Cupy):
+            cupy = ak.nplike.Cupy.instance()
+            return cupy.asarray(array)
+        elif isinstance(nplike, ak.nplike.Numpy):
+            numpy = ak.nplike.Numpy.instance()
+            return numpy.asarray(array)
+        elif isinstance(nplike, ak._v2._typetracer.TypeTracer):
+            return ak._v2._typetracer.TypeTracerArray(
+                dtype=array.dtype, shape=array.shape
+            )
+        else:
+            ak._v2._util.error(
+                TypeError(
+                    "Invalid nplike, choose between nplike.Numpy, nplike.Cupy, Typetracer or Jax",
+                )
+            )
+
+    def ascontiguousarray(self, array, dtype=None):
+        ak._v2._util.error(NotImplementedError)
+
+    def zeros(self, *args, **kwargs):
+        return self._module.zeros(*args, **kwargs)
+
+    def frombuffer(self, *args, **kwargs):
+        import numpy
+
+        np_array = numpy.frombuffer(*args, **kwargs)
+        return self._module.asarray(np_array)
+
+    def array_equal(self, array1, array2):
+        return self._module.array_equal(array1, array2)
+
+    def repeat(self, array, repeats, axis=None):
+        return self._module.repeat(array, repeats, axis)
+
+    def nan_to_num(self, *args, **kwargs):
+        return self._module.nan_to_num(*args, **kwargs)
```

Comment:
Calling `self._module.function_name(*args, **kwargs)` is what the superclass of all the nplikes does, so if you leave it unimplemented here, it will get that default implementation.

The `frombuffer` is a little different: by going through `numpy.frombuffer`, this will only work for JAX arrays on the CPU. Is it possible to test whether a Python buffer (a.k.a. `<memory>`) is in main memory or a GPU? Without importing CuPy?

Wait a minute: you know that `jax.numpy.frombuffer` exists, right?

Reply:
Oh I missed that function while writing this PR. I made the correction now.